genrl/agents/modelbased/cem/cem.py

```diff
@@ -0,0 +1,169 @@
+
+
+class CEM(ModelBasedAgent):
+    def __init__(
+        self,
+        *args,
+        network: str = "mlp",
+        policy_layers: tuple = (100,),
+        lr_policy=1e-3,
+        percentile: int = 70,
+        rollout_size,
+        **kwargs
+    ):
+        super(CEM, self).__init__(*args, **kwargs)
+        self.network = network
+        self.rollout_size = rollout_size
+        self.rollout = RolloutBuffer(self.rollout_size, self.env)
+        self.policy_layers = policy_layers
+        self.lr_policy = lr_policy
+        self.percentile = percentile
+
+        self._create_model()
+        self.empty_logs()
+
+    def _create_model(self):
+        self.state_dim, self.action_dim, discrete, action_lim = get_env_properties(
+            self.env, self.network
+        )
+        self.agent = get_model("p", self.network)(
+            self.state_dim,
+            self.action_dim,
+            self.policy_layers,
+            "V",
+            discrete,
+            action_lim,
+        )
+        self.optim = torch.optim.Adam(self.agent.parameters(), lr=self.lr_policy)
+
+    def plan(self):
+        state = self.env.reset()
+        self.rollout.reset()
+        states, actions = self.collect_rollouts(state)
+        return (states, actions, self.rewards[-1])
+
+    def select_elites(self, states_batch, actions_batch, rewards_batch):
+        reward_threshold = np.percentile(rewards_batch, self.percentile)
+        elite_states = [
+            s.unsqueeze(0).clone()
+            for i in range(len(states_batch))
+            if rewards_batch[i] >= reward_threshold
+            for s in states_batch[i]
+        ]
+        elite_actions = [
+            a.unsqueeze(0).clone()
+            for i in range(len(actions_batch))
+            if rewards_batch[i] >= reward_threshold
+            for a in actions_batch[i]
+        ]
+
+        return torch.cat(elite_states, dim=0), torch.cat(elite_actions, dim=0)
+
+    def select_action(self, state):
+        state = torch.as_tensor(state).float()
+        action, dist = self.agent.get_action(state)
+        return action, torch.zeros((1, self.env.n_envs)), dist.log_prob(action).cpu()
+
+    def update_params(self):
+        sess = [self.plan() for _ in range(100)]
+        batch_states, batch_actions, batch_rewards = zip(*sess)
+        elite_states, elite_actions = self.select_elites(
+            batch_states, batch_actions, batch_rewards
+        )
+        action_probs = self.agent.forward(elite_states.float())
+        loss = F.cross_entropy(
+            action_probs.view(-1, self.action_dim),
+            elite_actions.long().view(-1),
+        )
+        self.logs["crossentropy_loss"].append(loss.item())
+        loss.backward()
+        # torch.nn.utils.clip_grad_norm_(self.agent.parameters(), 0.5)
+        self.optim.step()
+
+    def get_traj_loss(self, values, dones):
+        # No need for this here
+        pass
+
+    def collect_rollouts(self, state: torch.Tensor):
```

Comment:
Looks pretty similar to the `OnPolicyAgent` method. Shouldn't this return `values` and `dones` though? Not sure if this is a consequence of the algo.